### PR TITLE
add: Machine Pool existence check 

### PR DESCRIPTION
**Pull request by VanillaSpoon**
This pr adds a machinePoolExists which checks if machine pools are present in the cluster. This is used to determine between the use of Machine Pools and Machine Sets.

This change also contains an updated namespace for the ocm secret.

### CODE REVIEW

File: controllers/appwrapper_controller.go [hunk 1]
```diff
@@ -223,28 +229,27 @@ func onAdd(obj interface{}) {
 	aw, ok := obj.(*arbv1.AppWrapper)
 	if ok {
 		klog.Infof("Found Appwrapper named %s that has status %v", aw.Name, aw.Status.State)
-		if aw.Status.State == arbv1.AppWrapperStateEnqueued || aw.Status.State == "" {
+		if aw.Status.State == arbv1.AppWrapperStateEnqueued || aw.Status.State == "" && aw.Labels != nil {
```
Comment (asm582) [inline controllers/appwrapper_controller.go:232]: appwrapper can have different kinds of labels, may be we should check if orderedInstance label exists or not

**Review by asm582 — COMMENTED** [review]
Thanks, Please address review

File: controllers/appwrapper_controller.go (hunk 1 above)
Comment (VanillaSpoon) [reply controllers/appwrapper_controller.go:232]: Thanks for pointing this out. 
Also that particular check is no longer needed due to #96

So, I think I'll remove that completely, what do you think?

Comment (asm582) [reply controllers/appwrapper_controller.go]: InstaScale looks for `"orderedinstance": "test.instance1"` to scale-up, so we should only scale-up when such a label exists inside an appwrapper, can you please confirm if we are doing the same?

File: controllers/appwrapper_controller.go [hunk 2]
```diff
@@ -149,27 +150,32 @@ func (r *AppWrapperReconciler) SetupWithManager(mgr ctrl.Manager) error {
...
-	useMachineSets = true
-	useMachinePools, err := strconv.ParseBool(instascaleConfigmap.Data["useMachinePools"])
-	if err != nil {
-		klog.Infof("Error converting %v to bool. Defaulting to using Machine Sets", useMachineSets)
+	if instascaleConfigmap != nil {
+		klog.Infof("Got config map named %v from namespace %v that configures max nodes in cluster to value %v", instascaleConfigmap.Name, instascaleConfigmap.Namespace, maxScaleNodesAllowed)
+	}
+	useMachineSets = false
+	// make an ocm call - if machine pools are available use them, if not default to machine sets
+	machinePoolExists := machinePoolExists()
```
Comment (dimakis) [inline controllers/appwrapper_controller.go:166]: This call will fail here if there is no OcmSecret provided right? we need the token to access OCM? 🤔

**Comment by dimakis** [comment]
The new changes are excellent. I would jus task that you let the error be a little more informative. maybe a `if looking to use ocm, the ocm secret must be available on cluster/in namespace` or something to that effect

File: controllers/appwrapper_controller.go [hunk 3]
```diff
@@ -149,36 +150,47 @@ func (r *AppWrapperReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	instascaleConfigmap, err := kubeClient.CoreV1().ConfigMaps(r.ConfigsNamespace).Get(context.Background(), "instascale-config", metav1.GetOptions{})
 	if err != nil {
 		klog.Infof("Config map named instascale-config is not available in namespace %v", r.ConfigsNamespace)
```
Comment (asm582) [inline controllers/appwrapper_controller.go:152]: Nit: log level should be error

File: controllers/appwrapper_controller.go [hunk 4]
```diff
@@ -149,36 +150,47 @@ func (r *AppWrapperReconciler) SetupWithManager(mgr ctrl.Manager) error {
...
+		return err
 	}
 
-	if maxScaleNodesAllowed, err = strconv.Atoi(instascaleConfigmap.Data["maxScaleoutAllowed"]); err != nil {
+	maxScaleNodesAllowed, err = strconv.Atoi(instascaleConfigmap.Data["maxScaleoutAllowed"])
+	if err != nil {
 		klog.Infof("Error converting %v to int. Setting maxScaleNodesAllowed to 3", maxScaleNodesAllowed)
```
Comment (asm582) [inline controllers/appwrapper_controller.go:158]: Nit: I think log level should be warn

**Review by asm582 — APPROVED** [review]
Please address Nit(s), thanks for your contribution

**Comment by VanillaSpoon** [comment]
Thanks for the feedback @asm582 and @dimakis :)

I hope these the recent changes reflect your suggestions accurately.

File: controllers/appwrapper_controller.go [hunk 5]
```diff
@@ -223,28 +237,27 @@ func onAdd(obj interface{}) {
...
-					scaleUp(aw, demandPerInstanceType)
-				} else {
-					klog.Infof("Cannot scale up replicas max replicas allowed is %v", maxScaleNodesAllowed)
+					if canScaleMachinepool(demandPerInstanceType) {
+						scaleUp(aw, demandPerInstanceType)
+					} else {
+						klog.Infof("Cannot scale up replicas max replicas allowed is %v", maxScaleNodesAllowed)
+					}
 				}
 			}
```
Comment (ChristianZaccaria) [inline controllers/appwrapper_controller.go:258]: I noticed there was a TODO for simplifying this loop. This reads a bit better:
```
if demandPerInstanceType != nil {
	if (useMachineSets && canScaleMachineset(demandPerInstanceType)) || (!useMachineSets && canScaleMachinepool(demandPerInstanceType)) {
		scaleUp(aw, demandPerInstanceType)
	} else {
		klog.Infof("Cannot scale up replicas. The maximum allowed replicas is %v", maxScaleNodesAllowed)
	}
}
```

**Review by ChristianZaccaria — CHANGES_REQUESTED** [review]

File: controllers/appwrapper_controller.go [hunk 6]
```diff
@@ -149,36 +150,49 @@ func (r *AppWrapperReconciler) SetupWithManager(mgr ctrl.Manager) error {
...
-	}
+	ocmSecretExists := ocmSecretExists(r.OcmSecretNamespace)
+	if ocmSecretExists {
+		machinePoolsExists := machinePoolExists()
 
-	if !useMachineSets {
-		instascaleOCMSecret, err := kubeClient.CoreV1().Secrets(r.ConfigsNamespace).Get(context.Background(), "instascale-ocm-secret", metav1.GetOptions{})
-		if err != nil {
-			klog.Errorf("Error getting instascale-ocm-secret from namespace %v - Error :  %v", r.ConfigsNamespace, err)
+		if machinePoolsExists && ocmSecretExists {
```
[the remark this answers is not in the document]
Comment (VanillaSpoon) [reply controllers/appwrapper_controller.go]: Great spot thanks @ChristianZaccaria

File: controllers/appwrapper_controller.go (hunk 5 above)
Comment (VanillaSpoon) [reply controllers/appwrapper_controller.go:258]: Thanks @ChristianZaccaria,
I'm working on a follow up pr with refactors, which will include this :)

Comment (ChristianZaccaria) [reply controllers/appwrapper_controller.go:258]: That sounds great, well done on this PR Eoin, looks fantastic :+1:

**Review by ChristianZaccaria — APPROVED** [review]

**Review by dimakis — APPROVED** [review]
LGTM great work @VanillaSpoon

**Pull request merged**